File: public/cpp/052-house-robber.cpp
/*
 * Problem: House Robber
 * LeetCode: https://leetcode.com/problems/house-robber/
 * 
 * Description:
 * You are a professional robber planning to rob houses along a street. Each house has a certain
 * amount of money stashed, the only constraint stopping you from robbing each of them is that
 * adjacent houses have security systems connected and it will automatically contact the police
 * if two adjacent houses were broken into on the same night.
 * 
 * Given an integer array nums representing the amount of money of each house, return the maximum
 * amount of money you can rob tonight without alerting the police.
 * 
 * Example 1:
 * Input: nums = [1,2,3,1]
 * Output: 4
 * Explanation: Rob house 1 (money = 1) and then rob house 3 (money = 3).
 * Total amount you can rob = 1 + 3 = 4.
 * 
 * Example 2:
 * Input: nums = [2,7,9,3,1]
 * Output: 12
 * Explanation: Rob house 1 (money = 2), rob house 3 (money = 9) and rob house 5 (money = 1).
 * Total amount you can rob = 2 + 9 + 1 = 12.
 * 
 * Constraints:
 * - 1 <= nums.length <= 100
 * - 0 <= nums[i] <= 400
 * 
 * Difficulty: Medium
 * Topics: Array, Dynamic Programming
 */

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    /*
     * Approach 1: Dynamic Programming (Bottom-Up with Array)
     * 
     * Intuition:
     * - At each house i, we have two choices:
     *   1. Rob house i: Take nums[i] + max money from houses up to i-2
     *   2. Skip house i: Take max money from houses up to i-1
     * - dp[i] = max(nums[i] + dp[i-2], dp[i-1])
     * 
     * Algorithm:
     * 1. Create dp array where dp[i] = max money robbing houses 0 to i
     * 2. Base cases: dp[0] = nums[0], dp[1] = max(nums[0], nums[1])
     * 3. For each house i from 2 to n-1:
     *    - dp[i] = max(nums[i] + dp[i-2], dp[i-1])
     * 4. Return dp[n-1]
     * 
     * Time Complexity: O(n) - single pass through array
     * Space Complexity: O(n) - dp array
     */
    int rob_dp_array(vector<int>& nums) {
        int n = nums.size();
        if (n == 0) return 0;
        if (n == 1) return nums[0];
        
        vector<int> dp(n);
        dp[0] = nums[0];
        dp[1] = max(nums[0], nums[1]);
        
        for (int i = 2; i < n; i++) {
            // Either rob current house + max from i-2, or skip and take max from i-1
            dp[i] = max(nums[i] + dp[i-2], dp[i-1]);
        }
        
        return dp[n-1];
    }
    
    /*
     * Approach 2: Dynamic Programming (Space Optimized)
     * 
     * Intuition:
     * - We only need the last two values (dp[i-1] and dp[i-2])
     * - Can optimize space from O(n) to O(1)
     * - Use two variables: prev1 (i-1) and prev2 (i-2)
     * 
     * Algorithm:
     * 1. Initialize prev2 = 0, prev1 = nums[0]
     * 2. For each house i from 1 to n-1:
     *    - current = max(nums[i] + prev2, prev1)
     *    - Update: prev2 = prev1, prev1 = current
     * 3. Return prev1
     * 
     * Time Complexity: O(n) - single pass
     * Space Complexity: O(1) - only two variables
     */
    int rob_optimized(vector<int>& nums) {
        int n = nums.size();
        if (n == 0) return 0;
        if (n == 1) return nums[0];
        
        int prev2 = 0;           // Max money up to i-2
        int prev1 = nums[0];     // Max money up to i-1
        
        for (int i = 1; i < n; i++) {
            int current = max(nums[i] + prev2, prev1);
            prev2 = prev1;
            prev1 = current;
        }
        
        return prev1;
    }
    
    /*
     * Approach 3: Recursion with Memoization (Top-Down DP)
     * 
     * Intuition:
     * - Recursive definition: rob(i) = max(nums[i] + rob(i-2), rob(i-1))
     * - Use memoization to avoid recomputing subproblems
     * - Cache results in memo array
     * 
     * Algorithm:
     * 1. Base cases: rob(0) = nums[0], rob(1) = max(nums[0], nums[1])
     * 2. For each house i:
     *    - If already computed, return memo[i]
     *    - Otherwise: memo[i] = max(nums[i] + rob(i-2), rob(i-1))
     * 3. Return rob(n-1)
     * 
     * Time Complexity: O(n) - each subproblem computed once
     * Space Complexity: O(n) - memo array + recursion stack
     */
    int robHelper(vector<int>& nums, int i, vector<int>& memo) {
        if (i < 0) return 0;
        if (i == 0) return nums[0];
        if (memo[i] != -1) return memo[i];
        
        // Rob current house + max from i-2, or skip and take max from i-1
        memo[i] = max(nums[i] + robHelper(nums, i-2, memo), 
                      robHelper(nums, i-1, memo));
        return memo[i];
    }
    
    int rob_memoization(vector<int>& nums) {
        int n = nums.size();
        if (n == 0) return 0;
        
        vector<int> memo(n, -1);
        return robHelper(nums, n-1, memo);
    }
    
    /*
     * Approach 4: Alternative Space Optimized (rob/skip pattern)
     * 
     * Intuition:
     * - Track two states: rob current house vs skip current house
     * - rob = nums[i] + skip (must skip previous if robbing current)
     * - skip = max(rob, skip) from previous iteration
     * 
     * Algorithm:
     * 1. Initialize rob = 0, skip = 0
     * 2. For each house:
     *    - newRob = skip + nums[i] (rob current, add to previous skip)
     *    - newSkip = max(rob, skip) (skip current, take max from previous)
     *    - Update rob = newRob, skip = newSkip
     * 3. Return max(rob, skip)
     * 
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     */
    int rob_rob_skip(vector<int>& nums) {
        int rob = 0;   // Max money if we rob current house
        int skip = 0;  // Max money if we skip current house
        
        for (int num : nums) {
            int newRob = skip + num;           // Rob current + previous skip
            int newSkip = max(rob, skip);      // Skip current, take max
            rob = newRob;
            skip = newSkip;
        }
        
        return max(rob, skip);
    }
    
    /*
     * Approach 5: Even More Concise Space Optimized
     * 
     * Intuition:
     * - Same as approach 2 but more concise
     * - Use prev and curr to track max money
     * 
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     */
    int rob(vector<int>& nums) {
        int prev = 0, curr = 0;
        
        for (int num : nums) {
            int temp = max(curr, prev + num);
            prev = curr;
            curr = temp;
        }
        
        return curr;
    }
};

/*
 * Test Cases
 */
void runTests() {
    Solution sol;
    
    // Test Case 1: Basic case
    vector<int> test1 = {1, 2, 3, 1};
    cout << "Test 1 - Input: [1,2,3,1]" << endl;
    cout << "DP Array: " << sol.rob_dp_array(test1) << endl;
    cout << "Optimized: " << sol.rob_optimized(test1) << endl;
    cout << "Memoization: " << sol.rob_memoization(test1) << endl;
    cout << "Rob/Skip: " << sol.rob_rob_skip(test1) << endl;
    cout << "Concise: " << sol.rob(test1) << endl;
    cout << "Expected: 4" << endl << endl;
    
    // Test Case 2: Longer array
    vector<int> test2 = {2, 7, 9, 3, 1};
    cout << "Test 2 - Input: [2,7,9,3,1]" << endl;
    cout << "DP Array: " << sol.rob_dp_array(test2) << endl;
    cout << "Optimized: " << sol.rob_optimized(test2) << endl;
    cout << "Memoization: " << sol.rob_memoization(test2) << endl;
    cout << "Rob/Skip: " << sol.rob_rob_skip(test2) << endl;
    cout << "Concise: " << sol.rob(test2) << endl;
    cout << "Expected: 12" << endl << endl;
    
    // Test Case 3: Single house
    vector<int> test3 = {5};
    cout << "Test 3 - Input: [5]" << endl;
    cout << "Result: " << sol.rob(test3) << endl;
    cout << "Expected: 5" << endl << endl;
    
    // Test Case 4: Two houses
    vector<int> test4 = {2, 1};
    cout << "Test 4 - Input: [2,1]" << endl;
    cout << "Result: " << sol.rob(test4) << endl;
    cout << "Expected: 2" << endl << endl;
    
    // Test Case 5: All same values
    vector<int> test5 = {5, 5, 5, 5, 5};
    cout << "Test 5 - Input: [5,5,5,5,5]" << endl;
    cout << "Result: " << sol.rob(test5) << endl;
    cout << "Expected: 15 (rob houses 0, 2, 4)" << endl << endl;
    
    // Test Case 6: Increasing values
    vector<int> test6 = {1, 2, 3, 4, 5};
    cout << "Test 6 - Input: [1,2,3,4,5]" << endl;
    cout << "Result: " << sol.rob(test6) << endl;
    cout << "Expected: 9 (rob houses 0, 2, 4 = 1+3+5)" << endl << endl;
    
    // Test Case 7: Large values at ends
    vector<int> test7 = {100, 1, 1, 100};
    cout << "Test 7 - Input: [100,1,1,100]" << endl;
    cout << "Result: " << sol.rob(test7) << endl;
    cout << "Expected: 200 (rob houses 0 and 3)" << endl << endl;
}

int main() {
    cout << "House Robber - Multiple Approaches" << endl;
    cout << "===================================" << endl << endl;
    
    runTests();
    
    return 0;
}

/*
 * Complexity Analysis Summary:
 * 
 * Approach 1 (DP Array):
 * - Time: O(n) - single pass through array
 * - Space: O(n) - dp array
 * 
 * Approach 2 (Space Optimized):
 * - Time: O(n) - single pass
 * - Space: O(1) - only two variables
 * - BEST for interviews: optimal time and space
 * 
 * Approach 3 (Memoization):
 * - Time: O(n) - each subproblem computed once
 * - Space: O(n) - memo array + recursion stack
 * 
 * Approach 4 (Rob/Skip):
 * - Time: O(n)
 * - Space: O(1)
 * - Alternative way to think about the problem
 * 
 * Approach 5 (Concise):
 * - Time: O(n)
 * - Space: O(1)
 * - Most concise implementation
 * 
 * Key Insights:
 * 1. Classic DP problem with optimal substructure
 * 2. Recurrence: dp[i] = max(nums[i] + dp[i-2], dp[i-1])
 * 3. Can optimize space from O(n) to O(1)
 * 4. Multiple ways to think about the problem (rob/skip, prev/curr)
 * 5. Base cases are crucial: handle n=0, n=1, n=2
 * 
 * Common Pitfalls:
 * 1. Forgetting to handle edge cases (empty array, single house)
 * 2. Off-by-one errors in indexing
 * 3. Not considering that skipping might be better than robbing
 * 4. Incorrect initialization of base cases
 * 
 * Interview Tips:
 * 1. Start with recursive solution to show understanding
 * 2. Identify overlapping subproblems
 * 3. Convert to bottom-up DP
 * 4. Optimize space if asked
 * 5. Discuss trade-offs between approaches
 * 6. Test with edge cases
 * 
 * Related Problems:
 * - House Robber II (circular array)
 * - House Robber III (binary tree)
 * - Delete and Earn
 * - Maximum Alternating Subsequence Sum
 */

// Made with Bob

File: src/data/tutorials/052-house-robber.js
/**
 * Tutorial content for #052 — House Robber
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `You are a robber planning to rob houses along a street. Each house has some amount of money. The only constraint: you cannot rob two adjacent houses (the alarm will go off). Return the maximum amount you can rob.`,
    example: `houses = [2, 7, 9, 3, 1]\n→ Rob house 0 (2) + house 2 (9) + house 4 (1) = 12\n→ Or rob house 1 (7) + house 3 (3) = 10\n✅ Answer: 12`,
    keyInsight: `At every house you face exactly one choice: rob it (and skip the next) or skip it. The best answer at each house depends only on the two results before it.`,
  },

  approaches: {
    'Recursive Memoization (Top-Down DP)': {
      intuition: `Start from house 0 and ask: "what's the best I can do from here?"\n\nIf I rob this house, I get its money — but then I must jump to house i+2.\nIf I skip this house, I move to house i+1 and try again.\n\nI pick whichever gives more money. But this creates overlapping subproblems — the same house index gets asked repeatedly. We fix that by caching (memoising) each answer once computed.`,
      steps: [
        `Create a memo array filled with -1 (means "not yet computed").`,
        `Call robHelper(nums, 0, memo) — start from house 0.`,
        `Base case: if index ≥ length, no houses left → return 0.`,
        `If memo[i] is already filled, return it immediately (cache hit).`,
        `Otherwise compute: max(nums[i] + helper(i+2), helper(i+1)).`,
        `Store the result in memo[i] before returning.`,
      ],
      example: `houses = [2, 7, 9, 3, 1]\n\nhelper(0): max(2 + helper(2), helper(1))\nhelper(2): max(9 + helper(4), helper(3))\nhelper(4): max(1 + helper(6), helper(5)) = 1\nhelper(3): max(3 + helper(5), helper(4)) = 3\nhelper(2) = max(9+1, 3) = 10  ← cached\nhelper(1): max(7 + helper(3), helper(2)) = max(10, 10) = 10\nhelper(0) = max(2+10, 10) = 12 ✅`,
      keyInsight: `Without memoisation this is O(2ⁿ). With it, each index is computed exactly once → O(n) time, O(n) space for the cache + recursion stack.`,
    },

    'Bottom-Up DP': {
      intuition: `Instead of starting at house 0 and recursing forward, we build the answer from the ground up.\n\nWe define dp[i] = the most money we can rob from houses 0 through i.\n\nFor each new house i we ask: is it better to rob it (dp[i-2] + nums[i]) or skip it (dp[i-1])? We just pick the max.`,
      steps: [
        `Handle the edge case: if only 1 house, return nums[0].`,
        `Set dp[0] = nums[0] (only one house available).`,
        `Set dp[1] = max(nums[0], nums[1]) (best of the first two).`,
        `For i from 2 to n-1: dp[i] = max(dp[i-1], dp[i-2] + nums[i]).`,
        `Return dp[n-1] — the answer considering all houses.`,
      ],
      example: `houses = [2, 7, 9, 3, 1]\n\ndp[0] = 2\ndp[1] = max(2, 7) = 7\ndp[2] = max(7, 2+9) = 11\ndp[3] = max(11, 7+3) = 11\ndp[4] = max(11, 11+1) = 12 ✅`,
      keyInsight: `The recurrence dp[i] = max(dp[i-1], dp[i-2] + nums[i]) is the heart of this problem. Every DP house-robber variant reduces to this.`,
    },

    'Space-Optimized DP': {
      intuition: `Look at the bottom-up recurrence: dp[i] only ever uses dp[i-1] and dp[i-2]. We don't need the whole array — just two variables.\n\nWe call them prev2 (two steps back) and prev1 (one step back), and slide them forward as we go.`,
      steps: [
        `Start with prev2 = 0, prev1 = 0 (no houses yet).`,
        `For each house value num:`,
        `  cur = max(prev1, prev2 + num)  ← rob or skip`,
        `  prev2 = prev1  ← slide the window`,
        `  prev1 = cur`,
        `Return prev1 — the answer after the last house.`,
      ],
      example: `houses = [2, 7, 9, 3, 1]\n\nprev2=0, prev1=0\nnum=2: cur=max(0, 0+2)=2   → prev2=0, prev1=2\nnum=7: cur=max(2, 0+7)=7   → prev2=2, prev1=7\nnum=9: cur=max(7, 2+9)=11  → prev2=7, prev1=11\nnum=3: cur=max(11, 7+3)=11 → prev2=11, prev1=11\nnum=1: cur=max(11,11+1)=12 → prev2=11, prev1=12\nReturn 12 ✅`,
      keyInsight: `This is the interview-optimal solution. Same O(n) time, but O(1) space — you only ever need to remember the previous two best values.`,
    },

    'Rob/Skip State Machine': {
      intuition: `Model the problem as two explicit states:\n• robVal = best total if you ROB the current house\n• skipVal = best total if you SKIP the current house\n\nAt each house, the new rob value is (previous skip + current house money), and the new skip value is the max of the previous two states.`,
      steps: [
        `Start with robVal = 0, skipVal = 0.`,
        `For each house num:`,
        `  newRob  = skipVal + num  ← can only rob after a skip`,
        `  newSkip = max(robVal, skipVal)  ← best of previous states`,
        `  Update robVal = newRob, skipVal = newSkip.`,
        `Return max(robVal, skipVal).`,
      ],
      example: `houses = [2, 7, 9, 3, 1]\n\nrob=0,  skip=0\nnum=2: newRob=0+2=2,  newSkip=max(0,0)=0  → rob=2,  skip=0\nnum=7: newRob=0+7=7,  newSkip=max(2,0)=2  → rob=7,  skip=2\nnum=9: newRob=2+9=11, newSkip=max(7,2)=7  → rob=11, skip=7\nnum=3: newRob=7+3=10, newSkip=max(11,7)=11 → rob=10, skip=11\nnum=1: newRob=11+1=12,newSkip=max(10,11)=11 → rob=12,skip=11\nReturn max(12, 11) = 12 ✅`,
      keyInsight: `The state-machine framing makes the "no two adjacent" constraint completely explicit — you can only transition from skip → rob, never rob → rob.`,
    },

    'Concise prev/curr': {
      intuition: `The most minimal way to write it. Two variables: prev and curr.\n\nAt each step, the new best is max(curr, prev + num). Then slide: prev becomes the old curr, curr becomes the new best.`,
      steps: [
        `Start with prev = 0, curr = 0.`,
        `For each house num:`,
        `  temp = max(curr, prev + num)`,
        `  prev = curr`,
        `  curr = temp`,
        `Return curr.`,
      ],
      example: `houses = [2, 7, 9, 3, 1]\n\nprev=0, curr=0\nnum=2: temp=max(0,0+2)=2   → prev=0, curr=2\nnum=7: temp=max(2,0+7)=7   → prev=2, curr=7\nnum=9: temp=max(7,2+9)=11  → prev=7, curr=11\nnum=3: temp=max(11,7+3)=11 → prev=11,curr=11\nnum=1: temp=max(11,11+1)=12→ prev=11,curr=12\nReturn 12 ✅`,
      keyInsight: `This is Space-Optimized DP with a different variable naming convention. Both are correct — pick whichever reads more naturally to you.`,
    },
  },
}
